src/hash.cr

```diff
@@ -36,12 +36,162 @@ class Hash(K, V)
+  # first (at this point computing the hash code plus comparing them might become
+  # cheaper than doing a full comparison each time). This optimization also exists
+  # in the Ruby implementation (though it seems hash values are always compared).
+  #
+  # A third optimization is in the way `@indices` is allocated: when the number
+  # of entries is less than or equal to 128 (2 ** 8 / 2) the indexes values will range between
+  # 0 and 128. That means we can use `Pointer(UInt8)` as the type of `@indices`.
+  # (we can't do it for ranges between 0 and 256 because we need a value that means
+  # "empty"). Similarly, for ranges between 128 and 32768 (2 ** 16 / 2) we can use
+  # `Pointer(UInt16)`. This saves some memory (and the performance difference is
+  # noticeable). We store the bytesize of the `@indices` buffer in `@indices_bytesize`
+  # with values 1 (UInt8), 2 (UInt16) or 4 (UInt32). This optimization also exists
+  # in the Ruby implementation.
+  #
+  # Another optimization is, when fitting a value inside the range of `@indices`,
+  # to use masking (value & mask) instead of `remainder` or `%`, which apparently
+  # are much slower. This optimization also exists in the Ruby implementation.
+  #
+  # We also keep track of the number of deleted entries (`@deleted_count`). When an
+  # entry is deleted we just mark it as deleted by using the special hash value 0.
+  # Only when the hash needs to be resized we do something with this instance variable:
+  # if we have many deleted entries (at least as many as the number of non-deleted
+  # entries) we compact the map and avoid a resize. Otherwise we remove the non-deleted
+  # entries but also resize both the `@entries` and `@indices` buffer. This probably
+  # avoids an edge case where one deletes and inserts an element and there is a constant
+  # shift of the buffer (expensive).
+  #
+  # There might be other optimizations to try out, like not using Linear Probing,
+  # but for now this implementaton is much faster than the old one which used
+  # linked lists (closed addressing).
+  #
+  # All methods that deal with this implementation come after the constructors.
+  # Then all other methods use the internal methods, usually using other high-level
+  # methods.
+
+  # The buffer of entries.
+  # Might be null if the hash is empty at the very beginning.
+  # Has always the size of `indices_size` / 2.
+  @entries : Pointer(Entry(K, V))
+
+  # The buffer of indices into entries. Its size is given by `@indices_size_pow2`.
+  # Might be null if the hash is empty at the very beginning or when the hash
+  # size is less than or equal to 16.
+  # Could be a Slice but this way we might save a few bounds checking.
+  @indices : Pointer(UInt8)
+
+  # The size of `@indices` given as a power of 2.
+  # For example if it's 4 it means 2**4 so size 16.
+  # Can be zero when hash is totally empty.
+  # Otherwise guaranteed to be at least 3.
+  @indices_size_pow2 : UInt32
```

Comment:
no need for `@indices_size_pow2` to be `UInt32`. It could be just `UInt8`. Then placed near `@indices_bytesize` it will be better fitting.

Reply:
You are right! I changed it to `UInt8`. I think previously I tracked `@indices_size` and that was `UInt32` and then I replaced it with pow2 but didn't realized the size could be slower.